keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1899 @@
+that specifies the assignment of pod-level QoS-class resources. The intended
+use for this would be to be able to communicate pod-level QoS-class resource
+assignments at sandbox creation time (`RunPodSandboxRequest`).
+
+```diff
+ message PodSandboxConfig {
+@@ -45,5 +45,14 @@ message PodSandboxConfig {
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++    // Configuration of QoS resources.
++    PodQoSResources qos_resources = 10;
+ }
+
++// PodQoSResources specifies the configuration of QoS resources of a pod.
++message PodQoSResources {
++    // QoS resources the pod will be assigned to.
++    // Key-value pairs where key is name of the QoS resource and value is the
++    // name of the class.
++    map<string, string> classes = 1;
++}
+```
+
+#### ContainerStatus
+
+The `ContainerResources` message (part of `ContainerStatus`) will be extended
+to report back QoS-class resource assignments of a container, similar to other
+resources.
+
+```diff
+@@ -1251,6 +1269,8 @@ message ContainerResources {
+     LinuxContainerResources linux = 1;
+     // Resource limits configuration specific to Windows container.
+     WindowsContainerResources windows = 2;
++    // Configuration of QoS resources.
++    ContainerQoSResources qos_resources = 3;
+```
+
+#### RuntimeStatus
+
+Extend the `RuntimeStatus` message with new `resources` field that is used to
+communicate the available QoS-class resources from the runtime to the client.
+This information is dynamic i.e. the available QoS-class resources (or their
+properties like available classes or per-class capacity) may change over time,
+e.g. after re-configuration of the runtime.
+
+Names of the QoS-class resources must be unique, also across pod-level and
+container-level resources. I.e. the same name must not be advertised as both a
+pod-level and container-level resource.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Information about the discovered resources
++    ResourcesInfo resources = 2;
++}
+
++// ResourcesInfo contains information about the resources discovered by the
++// runtime.
++message ResourcesInfo {
++    // Pod-level QoS resources available.
++    repeated QoSResourceInfo pod_qos_resources = 1;
++    // Container-level QoS resources available.
++    repeated QoSResourceInfo container_qos_resources = 2;
++}
+
++// QoSResourceInfo contains information about one type of QoS resource.
++message QoSResourceInfo {
++    // Name of the QoS resources. Name must be unique, also across pod and
++    // container level QoS resources.
++    string Name = 1;
++    // Mutable is set to true if this resource supports in-place updates i.e.
++    // the class of a running container or sandbox can be changed.
++    bool Mutable = 2;
```

Comment:
I find myself wondering: is this making the resource less opaque to kube? from kube's perspective aren't they all key values? 

going further: in a VPA scenerio, the autoscaler would need to know more intimately about these resources anyway (to know which to scale between, etc). What value would adding this give if we already need  to teach the autoscaler about the specifics of the resources themselves

Reply:
Yes, they are key-values. However, this information makes it possible to catch early (kube apiserver and kubelet) if a user is trying to update a QoS resource that doesn't support it (rather than failing late with an error at the runtime). Makes sense? Maybe the name "Mutable" should be something more descriptive.

I agree with the VPA scenario. It isn't generally applicable with QoS resources.

Reply:
ah okay I misinterpreted the reasoning for it

Reply:
My bad English 🙄 I'm open to suggestions how to update the comment to make it more evident (and prevent others falling in the same misinterpretation), I think I've tried to improve this once or twice but failed...